Rewrite installation readme for clarity; remove outdated references (e.g. CUDA on OSX) (#1595)
pytorch/botorch: README.md

- scipy
- multiple-dispatch

> ### Prerequisite only for MacOS users:

Reviewer: There shouldn't be a need to do this for Apple Silicon (unless the user wants to use MPS, but that's a separate story)
```suggestion
### Prerequisite only for MacOS users with Intel processors:
```

> Before installing BoTorch, we recommend first manually installing PyTorch, a required dependency of BoTorch. Installing it according to the [PyTorch installation instructions](https://pytorch.org/get-started/locally/) ensures that it is properly linked against MKL, a library that optimizes mathematical computation for Intel processors. This will result in up to an order-of-magnitude speed-up for Bayesian optimization, as at the moment, installing PyTorch from pip does not link against MKL.

Reviewer: We've generally tried to limit line lengths to around 100ish - mind doing this as well and adding a few breaks (here and below)? It's a lot easier to read in basic text editors/viewers that don't auto-wrap lines.

> Before installing BoTorch, we recommend first manually installing PyTorch, a required dependency of BoTorch. Installing it according to the [PyTorch installation instructions](https://pytorch.org/get-started/locally/) ensures that it is properly linked against MKL, a library that optimizes mathematical computation for Intel processors. This will result in up to an order-of-magnitude speed-up for Bayesian optimization, as at the moment, installing PyTorch from pip does not link against MKL.

Author: ```suggestion
Before installing BoTorch, we recommend first manually installing PyTorch, a required dependency of
BoTorch. Installing it according to the [PyTorch installation instructions](https://pytorch.org/get-started/locally/)
ensures that it is properly linked against MKL, a library that optimizes mathematical computation for Intel processors.
This will result in up to an order-of-magnitude speed-up for Bayesian optimization, as at the moment,
installing PyTorch from pip does not link against MKL.
```

The PyTorch installation instructions currently recommend:
1. Install [Anaconda](https://www.anaconda.com/distribution/#download-section). Note that there are different installers for Intel and M1 Macs.

> 2. Install PyTorch with `conda install pytorch torchvision -c pytorch`.

Reviewer: Do we want to add this here? This means that unless we monitor the PyTorch installation instructions this could easily go stale. Ideally we can avoid that by just pointing to the installation instructions. Wdyt?

> 2. Install PyTorch with `conda install pytorch torchvision -c pytorch`.

Author: ```suggestion
2. Install PyTorch following the [PyTorch installation instructions](https://pytorch.org/get-started/locally/).
Currently, this suggests running `conda install pytorch torchvision -c pytorch`.
```

The latest release of BoTorch is easily installed either via
[Anaconda](https://www.anaconda.com/distribution/#download-section) (recommended):

First install PyTorch:
```bash

> # Needed for OSX only, see below.

Reviewer: ```suggestion
# Needed for OSX on Intel Macs only, see below.
```

conda install pytorch torchvision -c pytorch 
```

> _Recommendation for MacOS users:_ PyTorch is a required dependency of BoTorch, and can be automatically installed via `pip`. However, we recommend installing PyTorch manually (as above) so that PyTorch is properly linked against MKL (a library that optimizes mathematical computation for Intel processors). This will result in up to an order-of-magnitude speed-up for Bayesian optimization, as at the moment, installing PyTorch from pip does not link against MKL.

Reviewer: ```suggestion
_Recommendation for MacOS users:_ PyTorch is a required dependency of BoTorch, and can be automatically installed via `pip`. However, if you are using an Intel Mac we recommend installing PyTorch manually (as above) so that PyTorch is properly linked against MKL (a library that optimizes mathematical computation for Intel processors). This will result in up to an order-of-magnitude speed-up for Bayesian optimization, as at the moment, installing PyTorch from pip does not link against MKL.
```

> If you need CUDA on MacOS, or want to customize your installation (i.e. CPU-only option), please follow the [PyTorch installation instructions](https://pytorch.org/get-started/locally/) to build from source.

Reviewer: So in more recent pytorch versions CUDA doesn't seem to be supported anymore on MacOS, but MPS (for Apple silicon) is: https://pytorch.org/get-started/locally/

We should update the readme here accordingly.

Author: Out of curiosity, has anyone tested BoTorch with MPS?

Reviewer: I believe @sdaulton has. From browsing through its docs, MPS current still seems to be in its infancy in terms of supporting accelerated linear algebra, but it looks like they're working on closing that gap.

Then install BoTorch:
```bash

> conda install botorch -c conda-forge

Reviewer: At the time the `conda-forge` channel was a fix to make sure `pyro-ppl` can be resolved (see https://github.com/pytorch/botorch/pull/1207). In general we have recommended people install gpytorch from the `gpytorch` and `botorch` from the `pytorch` channels, respectively, rather than relying on `conda-forge` (which may not have the latest packages).

However, if this copying of the package into `conda-forge` has been working reliably and most people use `conda-forge` anyway then we could consider just pointing to `conda-forge` to simplify the installation? 

```suggestion
conda install botorch -c pytorch -c gpytorch -c conda-forge
```

Author: yeah that seems right -- I thought that installing BoTorch from the pytorch channel rather than conda-forge fixed the MKL warning I was getting, but on second check I'm not sure that's right. I'll look into that more carefully and make sure I land on some install instructions that are actually working.

```

> or via `pip`:

Reviewer: This "or" is now pretty far away from the "either"
```suggestion
Alternatively, you can install BoTorch via `pip`:
```
